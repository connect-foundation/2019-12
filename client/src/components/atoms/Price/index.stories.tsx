import React from 'react';
import Price from '.';

export default {
  title: 'Atoms / Price',
};

export const index: React.FC = () => {
  return <Price separated={true}>{16000000}</Price>;
};
