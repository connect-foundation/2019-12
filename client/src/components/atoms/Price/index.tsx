import React from 'react';

import * as S from './style';

export interface PriceProps {
  mount: number;
  currency?: '₩' | '$';
  separated?: boolean;
}

export function currencyDecorator({ mount, currency, separated }: PriceProps) {
  const seperator = (targetStr: string) => {
    return targetStr
      .split('')
      .reverse()
      .reduce<string[]>((acc, cur, i, arr) => {
        if ((i + 1) % 3 === 0 && i !== arr.length - 1) {
          acc = [...acc, cur, ','];
          return acc;
        }
        acc.push(cur);
        return acc;
      }, [])
      .reverse()
      .join('');
  };

  let mountStr = `${mount}`;
  if (separated) {
    mountStr = seperator(mountStr);
  }

  return `${currency} ${mountStr}`;
}

function Price({
  mount,
  currency = '₩',
  separated = false,
}: PriceProps): React.ReactElement {
  const convertedCurrency = currencyDecorator({ mount, currency, separated });

  return <S.Wrapper>{convertedCurrency}</S.Wrapper>;
}

export default Price;
