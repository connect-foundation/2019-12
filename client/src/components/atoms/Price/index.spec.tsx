import React from 'react';
import { mount } from 'enzyme';

import Price, { currencyDecorator, PriceProps } from '.';

describe('Atom / Price', () => {
  describe('SNAPSHOT', () => {
    it('Rendering', () => {
      const wrapper = mount(
        <Price mount={16000000} currency={'₩'} separated={true} />,
      );
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('currencyDecorator function', () => {
    // given
    const targetData = {
      mount: 16000000,
      currency: '₩',
      separated: true,
    } as PriceProps;

    it('currency가 표시된다.', () => {
      // when
      const result = currencyDecorator(targetData);
      const currencyIndex = result.indexOf('₩');

      // then
      expect(currencyIndex).toBe(0);
    });

    it('separated에 따라 분리하거나 분리하지 않는다.', () => {
      // when
      const result = currencyDecorator(targetData);
      const commas = result.split(',');

      const firstCommaIndex = result.indexOf(',');
      const lastCommaIndex = result.indexOf(',', firstCommaIndex + 1);

      // then
      expect(commas).toHaveLength(3);
      expect(firstCommaIndex).toBe(4);
      expect(lastCommaIndex).toBe(8);
    });

    it('currency와 separated 된 값이 반환된다.', () => {
      // when
      const result = currencyDecorator(targetData);

      // then
      expect(result).toBe('₩ 16,000,000');
    });
  });
});
