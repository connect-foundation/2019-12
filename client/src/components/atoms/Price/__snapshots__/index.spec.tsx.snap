// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Atom / Price SNAPSHOT Rendering 1`] = `
<Price
  currency="₩"
  mount={16000000}
  separated={true}
>
  <styled.div>
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-bdVaJa",
            "isStatic": false,
            "lastClassName": "bDWFJH",
            "rules": Array [],
          },
          "displayName": "styled.div",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-bdVaJa",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
    >
      <div
        className=""
      >
        ₩ 16,000,000
      </div>
    </StyledComponent>
  </styled.div>
</Price>
`;
